### ISSUE TITLE
Need a way to secure (encrypt) our models

### ISSUE THREAD

**Issue by JohnCraigPublic** [microsoft/onnxruntime#3556]
Our product contains ONNX models and is a desktop PC product -- not in the cloud.  We need a way to prevent people just finding and taking our ONNX models.   Is there a way to encrypt or somehow protect our models?  If not, I think its important feature.

Using ONNX V5

**Describe the solution you'd like**
Encryption scheme, and hopefully one that does not slow down the loading (session creation) very much.

**Describe alternatives you've considered**
If I do it myself somehow, do I need to write the un-encrypted file onto disk in order to load it with the session constructor?  Or is there a way to do it all in memory?

**Additional context**
Running in C++ using VS2017  windows 10  OnnxRuntime latest version

**Comment by snnn** [comment]
Sorry it's not supported. Welcome to contribute.

**Comment by pfeatherstone** [comment]
You can use https://github.com/graphitemaster/incbin to embed the weights in your executable. Then you won't have any dangling files. But someone can figure out the offset of the file in your executable then read it.

**Comment by JohnCraigPublic** [comment]
I ended up writing my own encypter for ONNX models....

**Comment by alekk** [comment]
> I ended up writing my own encypter for ONNX models....

Dear John, would you mind sharing a few more details about the approach you have used to encrypt your ONNX models?

Thanks!

**Comment by JohnCraigPublic** [comment]
alekk -- the key thing to know is that there is a constructor defined for Session that takes a data buffer (for the model).  Previously, I had always used the constructor that takes a string path to the model file on disk.  But since a handy constructor exists using a data buffer, one can stream the (encrypted) file in, then unencrypt the buffer, then hand the buffer to the Session constructor.  This way, the model file written on disk (and subject to theft) is always an encrypted file.  Hope that helps.  If so, you owe me one!

**Comment by dlml** [comment]
@JohnCraigPublic  Dear John, would you mind sharing any code ?

**Comment by kleinicke** [comment]
I used pyAesCrypt to decrypt the model. Looked into the code of onnx.save_model and got the code I need and added the decryption to it. 
Let me sketch it for you. You might need other functions from the save_model when you have some special onnx models.
For me the `onnx.save_model` did nothing but:
```
proto= keras2onnx.convert_keras(...) //create onnx model
pString = proto.SerializeToString()
with open(cast(Text, filePath), "wb") as writable:
    writable.write(pString)
```
So I replaced the last two lines with
```        
fIn = io.BytesIO(pString)
with open(onnxPath.with_suffix(".onnx_enc8"), "wb") as fOut:
    pyAesCrypt.encryptStream(fIn, fOut, password, bufferSize)
```
How to encrypt and decrypt is explained here: https://github.com/marcobellaccini/pyAesCrypt

**Comment by skaiware** [comment]
> Sorry it's not supported. Welcome to contribute.

Hi @snnn 
I know the ticket is closed but the question for onnxrt to support encrypted models is still open.
Encrypting the onnx file could be done offline using whatever framework used for training.
What is missing is an option/addon for onnxrt to decrypt it using a given key.
I know it s possible to manually do it by first decrypting in ram and then loading the onnx Session from a mem buffer.
But this is adding an extra layer that would also have to be protected against hacking and that could be done by onnxrt internally.
It would just need :
- a AES decryt implementation : that could be provided by any header only lib as plusaes (see ref).
- a new onnxrt Sessions API method taking an extra param for the key
Any cons/would you accept a PR ?
Kind
WT.
Refs:
https://github.com/kkAyataka/plusaes
https://medium.com/@ovechko.056/tensorflow-pretrained-model-encryption-decryption-in-mobile-apps-e3e95209716a

**Comment by WilliamTambellini** [comment]
> Sorry it's not supported. Welcome to contribute.

Hi @snnn 
I know the ticket is closed but the question for onnxrt to support encrypted models is still open.
Encrypting the onnx file could be done offline using whatever framework used for training.
What is missing is an option/addon for onnxrt to decrypt it using a given key.
I know it s possible to manually do it by first decrypting in ram and then loading the onnx Session from a mem buffer.
But this is adding an extra layer that would also have to be protected against hacking and that could be done by onnxrt internally.
It would just need :
- a AES decryt implementation : that could be provided by any header only lib as plusaes (see ref).
- a new onnxrt Sessions API method taking an extra param for the key
Any cons/would you accept a PR ?
Kind
WT.
Refs:
https://github.com/kkAyataka/plusaes
https://medium.com/@ovechko.056/tensorflow-pretrained-model-encryption-decryption-in-mobile-apps-e3e95209716a

**Comment by snnn** [comment]
I feel people often get confused by two concepts: encryption and obfuscation. I feel what you want is obfuscation, which is for making reverse engineering difficult. And AES doesn't do any help on that.  For example, if you have a OAUTH2 API key and you want to keep it private to your application, you need to obfuscation your application so that people like me can't put a debugger on it.  There are a lot of commercial SDKs(e.g. VMProtect) to help this, and it is far beyond of onnxruntime's scope. 

What onnxruntime would do is adding confidential computing support. Like: https://azure.microsoft.com/en-us/solutions/confidential-compute/ .

**Comment by WilliamTambellini** [comment]
Hi @snnn 
On our side I confirm that we are in need of :  
- encrypting the onnx file on disk to prevent anyone to load/use it with any framework capable to import onnx files
- decrypt the onnx on the fly, given a key, at model loading time with onnxrt

For that use case, confidential computing/obfuscation would nt help much here, the priority being to make the onnx file that we deliver to our customers unusable without our software.
Another similar request here:
https://github.com/onnx/onnx/issues/3223#issuecomment-860333626

Kind regards
WT

**Comment by yaysummeriscoming** [comment]
> Hi @snnn
> On our side I confirm that we are in need of :
> 
> * encrypting the onnx file on disk to prevent anyone to load/use it with any framework capable to import onnx files
> * decrypt the onnx on the fly, given a key, at model loading time with onnxrt
> 
> For that use case, confidential computing/obfuscation would nt help much here, the priority being to make the onnx file that we deliver to our customers unusable without our software.
> Another similar request here:
> [onnx/onnx#3223 (comment)](https://github.com/onnx/onnx/issues/3223#issuecomment-860333626)
> 
> Kind regards
> WT

I'm in a similar situation. This feature would be really useful!

**Comment by dsj178** [comment]
> 

我们是否可以通过重命名或者混淆模型来进行模型加密

**Comment by dsj178** [comment]
> I feel people often get confused by two concepts: encryption and obfuscation. I feel what you want is obfuscation, which is for making reverse engineering difficult. And AES doesn't do any help on that. For example, if you have a OAUTH2 API key and you want to keep it private to your application, you need to obfuscation your application so that people like me can't put a debugger on it. There are a lot of commercial SDKs(e.g. VMProtect) to help this, and it is far beyond of onnxruntime's scope.
> 
> What onnxruntime would do is adding confidential computing support. Like: https://azure.microsoft.com/en-us/solutions/confidential-compute/ .
@snnn 
Can we implement model encryption by renaming or obfuscating OP? I think you should have a model encryption scheme！

**Comment by Bhuvanamitra** [comment]
> Hi @snnn On our side I confirm that we are in need of :
> 
> * encrypting the onnx file on disk to prevent anyone to load/use it with any framework capable to import onnx files
> * decrypt the onnx on the fly, given a key, at model loading time with onnxrt
> 
> For that use case, confidential computing/obfuscation would nt help much here, the priority being to make the onnx file that we deliver to our customers unusable without our software. Another similar request here: [onnx/onnx#3223 (comment)](https://github.com/onnx/onnx/issues/3223#issuecomment-860333626)
> 
> Kind regards WT

I am in need of this feature. To deploy onnx models in On-Premise solutions, providing this feature in-built will be a huge advantage.

**Comment by monkeycc** [comment]
https://github.com/yangxianpku/model_cryptor

```
@unique
class ModelType(Enum): 
    MODEL_TORCH            = "TorchModelCryptor"            # torch模型
    MODEL_TORCHSCRIPT      = "TorchScriptModelCryptor"      # torch script模型, torch.jit.script或torch.jit.trace保存
    MODEL_ONNX             = "ONNXModelCryptor"             # onnx模型
    MODEL_TENSORRT         = "TensorRTModelCryptor"         # tensorrt模型
    MODEL_TORCH2TRT        = "Torch2TRTModelCryptor"        # torch2trt模型
    MODEL_TENSORFLOW       = "TensorFlowModelCryptor"       # tensorflow模型
    MODEL_TF2TRT           = "TF2TRTModelCryptor"           # tensorflow2tensorrt模型
    MODEL_PADDLE           = "PaddleModelCryptor"           # paddlepaddle模型
    MODEL_PADDLE2TRT       = "Paddle2TRTModelCryptor"       # paddle2tensorrt模型
```

**Comment by WilliamTambellini** [comment]
This is a humble request to at least make onnxrt API more friendly with encrypted models.

The onnxruntime/core/session/experimental_onnxruntime_cxx_api.h
offers 2 API:
```
  Session(Env& env, std::basic_string<ORTCHAR_T>& model_path, SessionOptions& options)
      : Ort::Session(env, model_path.data(), options){};

  Session(Env& env, void* model_data, size_t model_data_length, SessionOptions& options)
      : Ort::Session(env, model_data, model_data_length, options){};

```
The 2nd one allows the user to open his encrypted onnx model manually (file), read and decrypt it fully in mem, and then load from mem using the 2nd API (void* model).
The main issue is still that the decrypted model must then be entirely in mem while onnxrt loads it, meaning the model to be 'twice' in mem at the end of onnxrt loading, before the user could free his decrypted mem buffer.
 @snnn  do you understand the issue?
Best regards
